src/argilla_server/contexts/bulk/records/records_create_bulk.py

```diff
@@ -0,0 +1,71 @@
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import asyncio
+from typing import List
+
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from argilla_server.contexts.bulk.records import helpers
+from argilla_server.models import Dataset, Record
+from argilla_server.schemas.v1.records import RecordsCreate
+from argilla_server.search_engine import SearchEngine
+from argilla_server.validators.records import RecordCreateValidator
+
+
+class RecordsCreateBulk:
+
+    def __init__(self, db: AsyncSession, search_engine: SearchEngine):
+        self._db = db
+        self._search_engine = search_engine
+
+    async def create_dataset_records(self, dataset: Dataset, records_create: RecordsCreate) -> List[Record]:
+        helpers.check_is_ready_dataset(dataset)
+
+        records = []
+        async with self._db.begin_nested():
+            for idx, record_create in enumerate(records_create.items):
+                try:
+                    RecordCreateValidator(record_create).validate_for(dataset)
+                    record = Record(
+                        fields=record_create.fields,
+                        metadata_=record_create.metadata,
+                        external_id=record_create.external_id,
+                        dataset_id=dataset.id,
+                    )
+                    records.append(record)
+                except ValueError as ex:
+                    raise ValueError(f"Record at position {idx} is not valid because {ex}") from ex
+
+            self._db.add_all(records)
+            await self._db.flush(records)
+
+            await self._create_records_relationships(records, records_create)
+            await helpers.preload_records_relationships_before_index(self._db, records)
+            await self._search_engine.index_records(dataset, records)
+
+        await self._db.commit()
+        return records
+
+    async def _create_records_relationships(self, records: List[Record], records_create: RecordsCreate) -> None:
+
+        records_and_suggestions = list(zip(records, [r.suggestions for r in records_create.items]))
+        records_and_responses = list(zip(records, [r.responses for r in records_create.items]))
+        records_and_vectors = list(zip(records, [r.vectors for r in records_create.items]))
+
+        await asyncio.gather(
```

Comment:
Did you take a look to how the function work with exceptions? The default parameter `return_exceptions=False` is working like described:
> If return_exceptions is False (default), the first raised exception is immediately propagated to the task that awaits on gather(). Other awaitables in the aws sequence won’t be cancelled and will continue to run.
> If return_exceptions is True, exceptions are treated the same as successful results, and aggregated in the result list.


I believe we are safe because this inside a database transaction and the first exception will rollback everything but in the case that you want to take a look to the documentation: https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently

Reply:
Yes, the default value works for us.